base/shell/explorer/syspager.cpp

```diff
@@ -178,6 +178,90 @@ class CNotifyToolbar :
     void Initialize(HWND hWndParent, CBalloonQueue * queue);
 };
 
+
+class CHardErrorThread
+{
+    DWORD m_ThreadId;
+    LONG m_bThreadRunning;
+    DWORD m_Status;
+    DWORD m_dwResponseType;
+    CStringW m_Title;
+    CStringW m_Text;
+public:
+
+    CHardErrorThread():
+        m_bThreadRunning(FALSE),
+        m_Status(NULL),
+        m_dwResponseType(NULL),
+        m_ThreadId(NULL)
+    {
+    }
+
+    ~CHardErrorThread()
+    {
+        if (m_bThreadRunning)
+        {
+            /* Try to unstuck Show */
+            PostThreadMessage(m_ThreadId, WM_QUIT, 0, 0);   
+
+            HANDLE hThread = OpenThread(THREAD_ALL_ACCESS, FALSE, m_ThreadId);
+            if (hThread)
+            {
+                WaitForSingleObject(hThread, INFINITE);
+                CloseHandle(hThread);
+            }
+        }
+    }
+
+    VOID ThreadProc()
+    {
+        HRESULT hr;
+        CComPtr<IUserNotification> pnotification;
+
+        OleInitialize(NULL);
+
+        hr = CoCreateInstance(CLSID_UserNotification,
+                              NULL,
+                              CLSCTX_INPROC_SERVER,
+                              IID_PPV_ARG(IUserNotification, &pnotification));
+        if (SUCCEEDED(hr))
+        {
+            hr = pnotification->SetBalloonInfo(m_Title, m_Text, NIIF_WARNING);
```

Comment:
For the icon, do you think we should always keep the Warning icon (as Windows does), or, may we actually use one according to the Severity of the Status error of the notification?

Reply:
I want to test it more to see if there is a way to use a different icon.

Reply:
Testing in what sense? Yes of course there is a way to use different icons.
You check the Severity value of the Status error code (check the winsrv code) for informational, warning or error severity.
Then for the notification, you call first the balloon info as you did, but with a different NIIF_xxx value, then you call SetIconInfo with a NULL icon handle (and some tooltip text). When this order of operations is done, SetIconInfo detects that the icon handle is NULL in which case, it'll use an icon corresponding to the one the balloon will use.

Reply:
Testing in windows to see what parameters make it use different icons. Preliminary testing showed that it can't.

Reply:
You mean, the balloon COM interface in general, or the hard error notifications in particular?
If you talk about the harderror notifications in particular, they only have the warning icon because they hardcode the NIIF_WARNING flag. But in general, the IUserNotification interface allows one to change the icon as I described above, so we can use that to have a different icon depending on the "severity" of the status code.

Reply:
I mean testing in windows passing different params to NtRaiseHardError and see wtf is shown. I know what IUserNotification can do. What I don't know is what NtRaiseHardError can do.